# Homework no.2
**Student name:** Mohammad Amin Dadgar <br>
**Student Id:** 4003624016

In [1]:
import numpy as np
import matplotlib.pyplot as plt

# Q1
Using the training set below answer these question.
\begin{equation}
\text{Training Set }= \{(0.5, 2), (0.7, 3),(1, 8), (1.5, 10), (2, 10.3), (2.7, 5.5), (3, 8), (3.3, 11), (4, 10.2), (4.3, 12)\}
\end{equation}
Each data is $(x_i, y_i)$

## P1
fit the model $y=ax^2+bx+c$ on the given dataset and name the model as model_1. This means
\begin{equation}
\phi_1(x) = 
\begin{bmatrix}
1 \\
x \\
x^2 \\
\end{bmatrix}
\end{equation}
And the weights matrix is
\begin{equation}
W_1 = 
\begin{bmatrix}
c \\
b \\
a \\
\end{bmatrix}
\end{equation}

In [2]:
def model_1(training_set):
    """
    Linear Regression using quadratic function as the kernel function

    
    Parameters:
    -----------
    training_set : array of tuples
        array of `x_i` and `y_i`
        `x` the input variables
        `y` are the output variables
        
    Returns:
    ----------
    w : matrix_like
        the matrix of learned weights
    """
    ## Extract the X values and covert them using kernel function
    X = np.array(training_set)[:, 0]
    X = np.matrix(X).T
    
    ## apply the quadratic function to input data
    X_phi = kernel_Quadratic(X)
    ## extract the outputs
    Y = np.array(training_set)[:, 1]
    Y = np.matrix(Y)

    ## the `A` and `b` parameters of a Linear Regression function
    A = X_phi @ X_phi.T
    b = np.multiply(Y, X_phi)
    b = np.sum(b, axis=1)
    
    w = np.dot(np.linalg.inv(A), b)
    return w
    
    
def kernel_Quadratic(X):
    """
    The quadratic kernel function
    
    Parameters:
    -----------
    X : array_like
        the variables to be converted
   
    Returns:
    --------
    phi : array_like
        the quadratic function applied to `X`
    """
    ## first convert to a numpy array and copy it
    X_new = np.array(X).copy()
    
    ones_arr = np.ones(len(X))
    ones_arr = np.matrix(ones_arr)
    
    X_power_2 = np.power(X_new, 2)

    phi = np.vstack((ones_arr, X_new.T, X_power_2.T))
    
    return phi

In [3]:
train = [(0.5,2),(0.7,3),(1,8),(1.5,10),(2,10.3),(2.7,5.5),(3,8),(3.3,11),(4,10.2),(4.3,12)]
train = np.array(train)

In [4]:
model1_weight = model_1(train)
model1_weight

matrix([[ 1.94132333],
        [ 4.13814418],
        [-0.49656258]])

## P2
fit the model $y=ax^2+bx+c$ on the given dataset and name the model as model_1. This means
\begin{equation}
\phi_2(x) = 
\begin{bmatrix}
1 \\
x \\
x^2 \\
x^3 \\
\end{bmatrix}
\end{equation}
And the weights matrix is
\begin{equation}
W_2 = 
\begin{bmatrix}
d \\
c \\
b \\
a \\
\end{bmatrix}
\end{equation}

In [5]:
def mode_2(training_set):
    """
    Linear Regression function using ```ax^3 + bx^2 + cx + d``` as kernel function
    
    Parameters:
    -----------
    training_set : array of tuples
        array of `x_i` and `y_i`
        `x` the input variables
        `y` are the output variables
        
    Returns:
    ----------
    w : matrix_like
        the matrix of learned weights
    """

    ## Extract the X values and covert them using kernel function
    X = np.array(training_set)[:, 0]
    X = np.matrix(X).T
    
    ## apply the quadratic function to input data
    X_phi = kernel_cubic(X)
    ## extract the outputs
    Y = np.array(training_set)[:, 1]
    Y = np.matrix(Y)

    ## the `A` and `b` parameters of a Linear Regression function
    A = X_phi @ X_phi.T
    b = np.multiply(Y, X_phi)
    b = np.sum(b, axis=1)
    
    w = np.dot(np.linalg.inv(A), b)
    return w

def kernel_cubic(X):
    """
    The cubic function
    
    
    Parameters:
    -----------
    X : array_like
        the variables to be converted
        
    Returns:
    --------
    phi : array_like
        the quadratic function applied to `X`
    """
    
    ## first convert to a numpy array and copy it
    X_new = np.array(X).copy()
    
    ones_arr = np.ones(len(X))
    ones_arr = np.matrix(ones_arr)
    
    X_power_2 = np.power(X_new, 2)
    X_power_3 = np.power(X_new, 3)

    phi = np.vstack((ones_arr, X_new.T, X_power_2.T, X_power_3.T))
    
    return phi

In [6]:
model2_weight = mode_2(train)
model2_weight

matrix([[-7.33397387],
        [22.63114975],
        [-9.53106286],
        [ 1.24181752]])

## P3
Calculate the MSE for $\text{model}_1$ and $\text{model}_2$ using the dataset below
\begin{equation}
\text{Test set }= \{(3.8, 12.9), (1, 9),(2, 8), (1.9, 9), (1.5, 7), (1.6, 9.5), (1.2, 10), (3.3, 11), (1.5, 11), (4, 6)\}
\end{equation}
\end{equation}

In [14]:
test = np.array(((3.8, 12.9), (1, 9),(2, 8), (1.9, 9), 
                    (1.5, 7), (1.6, 9.5), (1.2, 10), (3.3, 11),
                    (1.5, 11), (4, 6)))
test

array([[ 3.8, 12.9],
       [ 1. ,  9. ],
       [ 2. ,  8. ],
       [ 1.9,  9. ],
       [ 1.5,  7. ],
       [ 1.6,  9.5],
       [ 1.2, 10. ],
       [ 3.3, 11. ],
       [ 1.5, 11. ],
       [ 4. ,  6. ]])

In [8]:
## applying model_1
def model_predict(weights, data, **kwargs):
    """
    predict the output of a model in this question 
    using the trained weights for the data
    
    the desired model can be selected using the quadratic or the cubic parameter
    
    Parameters:
    ------------
    weights : array_like
        the learned parameters for model1
    data : array or scaler
        the desired data for prediction task
    kwargs : 
        quadratics : use the model_1 in this question 
        cubic : use the model_2 in this question
        
    **NOTE:** one of the parameters can be choosed!
    **NOTE:** any value can be passed to kwargs, there is no difference in that. (because the value would not be used)
    
    Returns:
    --------
    prediction : array or scaler
        the predicted output value for the input data
    """
    
    ## first we need to apply the transformation for the data 
    ## as the learned parameters are trained using the transformation
    if 'quadratics' in kwargs and 'cubic' in kwargs:
        raise ValueError("Both parameters cannot be chosen! They have different behaviour")
    elif 'quadratics' in kwargs:
        data_phi = kernel_Quadratic(data) 
    elif 'cubic' in kwargs:
        data_phi = kernel_cubic(data) 
    else:
        raise ValueError("no parameters cubic or quadratics was chosen!")
        
    prediction = np.matrix(weights).T @ data_phi
    
    return prediction

In [12]:
def calculate_MSE(true_outputs, predicted_outputs):
    """
    calculate the Mean Squared Error for the predicted data
    
    Parameters:
    ------------
    true_outputs : array_like 
        the actual output of the data
    predicted_outputs : array_like
        the predicted outputs for the data
        
    Returns:
    ---------
    mse_value : float
        the mean squared error value
    """
    
    ## the difference 
    diff = predicted_outputs - true_outputs
    ## squeare the difference
    diff = np.power(diff, 2)
    
    ## find the mean value
    mse_value = np.mean(diff)
    
    return mse_value
    

In [15]:
pred_model1_test = model_predict(model1_weight, test[:, 0], quadratics='True')
pred_model2_test = model_predict(model2_weight, test[:, 0], cubic='True')

In [16]:
mse_model1 = calculate_MSE(test[:, 1], pred_model1_test)
mse_model2 = calculate_MSE(test[:, 1], pred_model2_test)
print(f'MSE for model1 is {mse_model1}')
print(f'MSE for model2 is {mse_model2}')

MSE for model1 is 7.496789449073051
MSE for model2 is 5.812331603917104


## P4

To find the T value for our datasets we need to use the equation below
\begin{equation}
T = \frac{mean_1 - mean_2}{\frac{std(diffrences)}{\sqrt{n}}}
\end{equation}
refrence: https://www.investopedia.com/terms/t/t-test.asp

In [17]:
def calculate_T_value(dataset1, dataset2):
    """
    Find the T-value for two datasets
    
    Parameters:
    ------------
    dataset1 : array_of_Tuples
        (`x_i`, `y_i`) data, `x` representing data and `y` representig label
    dataset2 : array_of_Tuples
        (`x_i`, `y_i`) data, `x` representing data and `y` representig label

    Returns:
    ---------
    T_value : float
        the T value for datasets
    """
    ## convert to numpy array for ease of use
    set1 = np.array(dataset1)
    set2 = np.array(dataset2)
    
    ## calculate the medians
    m1 = np.mean(set1[:, 1])
    m2 = np.mean(set2[:, 1])
    
    ## find the differences
    diff = set1[:, 1] - set2[:, 1]
    
    ## below the division
    below = np.std(diff) / np.sqrt(len(diff))
    
    above = m1 - m2
    
    T_value = above / below
    
    return T_value

In [20]:
calculate_T_value(train, test)

-0.9350212787977713

In [28]:
z = (train[:, 1] - np.mean(train[:, 1]- test[:, 1]) ) / np.std(train[:, 1] - test[:, 1])

In [30]:
np.mean(z)

2.0609315717988514

TODO: As we can see the value for T-test is about `-0.935` and it is not out of the [-1.96, 1.96]. So we can say that the 